firebase-firestore/src/main/java/com/google/firebase/firestore/local/OverlayedDocument.java

```diff
@@ -22,21 +22,27 @@
 public class OverlayedDocument {
   private Document overlayedDocument;
   private FieldMask mutatedFields;
+  private boolean hasOverlay;
 
-  OverlayedDocument(Document overlayedDocument, FieldMask mutatedFields) {
+  OverlayedDocument(Document overlayedDocument, boolean hasOverlay, FieldMask mutatedFields) {
     this.overlayedDocument = overlayedDocument;
     this.mutatedFields = mutatedFields;
+    this.hasOverlay = hasOverlay;
   }
 
   public Document getDocument() {
     return overlayedDocument;
   }
 
   /**
-   * The fields that are locally mutated by patch mutations. If the overlayed document is from set
-   * or delete mutations, this returns null.
+   * The fields that are locally mutated by patch mutations.
+   *
+   * <p>If hasOverlay() is true and the overlayed document is from set or delete mutations, this
+   * returns null. If hasOverlay() is false, this returns FieldMask.EMPTY.
    */
+  // TODO(b/262245989): This screams for a proper sum type (Tagged Union) which does not exist in
+  // Java (yet).
   public @Nullable FieldMask getMutatedFields() {
-    return mutatedFields;
+    return hasOverlay ? mutatedFields : FieldMask.EMPTY;
```

Comment:
Since `mutatedFields` is Nullable, I'm thinking whether it should be the responsibility of the caller (who is creating the OverlayedDocument object) to pass `null` for `mutatedFields` when we have a set/delete mutation, and pass `EMPTY` when there's no mutation. This way, this class won't need to store the `hasOverlay` boolean nor does it have to have this logic.

Reply:
Done.